packages/lib/src/delivery/Messages.ts

```diff
@@ -37,29 +47,82 @@ export async function getMessages(
+    deliveryServiceProfile: DeliveryServiceProfile,
     getSession: (accountAddress: string) => Promise<Session | null>,
     storeNewMessage: (
         conversationId: string,
         envelop: EncryptionEnvelop,
     ) => Promise<void>,
     send: (socketId: string, envelop: EncryptionEnvelop) => void,
 ): Promise<void> {
-    const envelop = {
-        ...data.envelop,
-        deliveryServiceIncommingTimestamp: new Date().getTime(),
-    };
-    const account = formatAddress(formatAddress(data.envelop.from));
-    const contact = formatAddress(formatAddress(data.envelop.to));
-    const conversationId = getConversationId(account, contact);
-
-    if (await checkToken(getSession, account, data.token)) {
-        storeNewMessage(conversationId, envelop);
-        const contactSession = await getSession(contact);
-        if (contactSession?.socketId) {
-            send(contactSession.socketId, envelop);
-        }
-    } else {
+    const sender = formatAddress(envelop.from);
+    const receiver = formatAddress(envelop.to);
+    const conversationId = getConversationId(sender, receiver);
+
+    const tokenIsValid = await checkToken(getSession, sender, token);
+    if (!tokenIsValid) {
+        //Token is invalid
         throw Error('Token check failed');
     }
+
+    const receiverSession = await getSession(receiver);
+    if (receiverSession === null) {
+        //TODO how should we handle this case?
+        throw Error('unknown session');
+    }
+
+    const receiverEncryptionKey =
+        receiverSession?.signedUserProfile.profile.publicEncryptionKey;
+
+    const deliveryServiceSigningKey = deliveryServiceProfile.publicSigningKey;
+
+    const envelopWithPostmark: EncryptionEnvelop = {
+        ...envelop,
+        postmark: addPostmark(
+            envelop,
+            receiverEncryptionKey,
+            deliveryServiceSigningKey,
+        ),
+    };
+    await storeNewMessage(conversationId, envelopWithPostmark);
+
+    if (receiverSession?.socketId) {
+        //Client is already connect to the delivery service and the message can be dispatched
+        send(receiverSession.socketId, envelopWithPostmark);
+    }
 }
+
+function addPostmark(
+    { encryptedData }: EncryptionEnvelop,
+    receiverEncryptionKey: string,
+    deliveryServiceSigningKey: string,
+): string {
+    const postmarkWithoutSig: Omit<Postmark, 'signature'> = {
+        messageHash: ethers.utils.hashMessage(encryptedData),
+        incommingTimestamp: new Date().getTime(),
+    };
+
+    const signature = signPostmark(
+        postmarkWithoutSig,
+        deliveryServiceSigningKey,
+    );
+
+    //Encrypte the signed Postmark and return the ciphertext
+    const { ciphertext, nonce, version, ephemPublicKey } = encryptSafely({
+        publicKey: receiverEncryptionKey,
+        data: { ...postmarkWithoutSig, signature },
+        version: 'x25519-xsalsa20-poly1305',
+    });
+
+    return stringify({ nonce, version, ciphertext, ephemPublicKey })!;
```

Comment:
I just returned the stringified encrypted object. Should be further encoded i.E Base64 ?

Reply:
I think we should discuss this on a call

Reply:
We let it unencoded